Android/app/src/main/java/app/intra/socks/SocksVpnAdapter.java

```diff
@@ -0,0 +1,177 @@
+package app.intra.socks;
+
+import com.google.firebase.crash.FirebaseCrash;
+
+import android.annotation.TargetApi;
+import android.os.Build;
+import android.os.ParcelFileDescriptor;
+
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.Locale;
+
+import androidx.annotation.NonNull;
+import app.intra.DnsVpnService;
+import app.intra.LocalhostResolver;
+import app.intra.VpnAdapter;
+import app.intra.util.LogWrapper;
+import app.intra.util.SafeTun2Socks;
+import sockslib.common.methods.NoAuthenticationRequiredMethod;
+
+/**
+ * This is a VpnAdapter that captures all traffic and routes it through a local SOCKS server.
+ */
+public class SocksVpnAdapter extends VpnAdapter {
+  private static final String LOG_TAG = "SocksVpnAdapter";
+
+  // IPv4 VPN constants
+  private static final String IPV4_TEMPLATE = "10.111.222.%d";
+  private static final int IPV4_PREFIX_LENGTH = 24;
+  private static final String IPV4_NETMASK = "255.255.255.0";
+
+  // Randomly generated unique local IPv6 unicast subnet prefix, as defined by RFC 4193.
+  private static final String IPV6_TEMPLATE = "fd66:f83a:c650::%d";
+  private static final int IPV6_PREFIX_LENGTH = 120;
+
+  private static final int MTU = 32767;
+  // Disable Transparent DNS.  We need Transparent DNS to be disabled because it is not compatible
+  // with standard SOCKS-UDP servers, only special servers that use a constant UDP port.
+  private static final int TRANSPARENT_DNS_ENABLED = 0;
+  private static final int SOCKS5_UDP_ENABLED = 1;
+
+  // The VPN service and tun2socks must agree on the layout of the network.  By convention, we
+  // assign the following values to the final byte of an address within a subnet.
+  private enum LanIp {
+    GATEWAY(1), ROUTER(2), DNS(3);
+
+    // Value of the final byte, to be substituted into the template.
+    private final int value;
+
+    LanIp(int value) {
+      this.value = value;
+    }
+
+    String make(String template) {
+      return String.format(Locale.ROOT, template, value);
+    }
+  }
+
+  // DNS resolver running on localhost.
+  private LocalhostResolver resolver;
+
+  // TUN device representing the VPN.
+  private ParcelFileDescriptor tunFd;
+
+  public static SocksVpnAdapter get(@NonNull DnsVpnService vpnService) {
+    LocalhostResolver resolver = LocalhostResolver.get(vpnService);
+    if (resolver == null) {
+      return null;
+    }
+    ParcelFileDescriptor tunFd = establishVpn(vpnService);
+    if (tunFd == null) {
+      return null;
+    }
+    return new SocksVpnAdapter(tunFd, resolver);
+  }
+
+  private SocksVpnAdapter(ParcelFileDescriptor tunFd, LocalhostResolver resolver) {
+    super(LOG_TAG);
+    this.resolver = resolver;
+    this.tunFd = tunFd;
+  }
+
+  @Override
+  public void run() {
+    resolver.start();
+
+    // VPN parameters
+    final int DNS_PORT = 53;
```

Comment:
This constant is also used in `DnsVpnAdapter`.

Reply:
Moved to `VpnAdapter`.